### ISSUE TITLE
Memory pre-loading

### ISSUE THREAD

**Issue by Al-Kindi-0** [maticnetwork/miden#336]
The following issue came up during discussions related to the implementation of the number theoretic transform (NTT). 
In many situations, it is required to have arrays of large sizes loaded into memory before the actual program starts running. This is the situation for example when doing NTT where a large array containing roots of unity is needed during the iterations of the algorithm. Clearly these constants should reside in memory as they are, most of the time, used several times during the execution of any useful program.
The question that poses itself then is how to do this in the most efficient+secure way. Some ideas that came up during the discussion:

1. One might augment the current mechanism used to load advice inputs in order to target the above situations in an efficient manner.
2. The memory trace can be pre-populated with the data by the prover before the start of the execution. This should be better, I believe, than the previous proposal but it is not immediately clear, to me at least, how we can insure soundness with this modification  while still using the methodology we currently use for insuring soundness i.e. without any additional permutation-checks machinery.

**Comment by bobbinth** [comment]
I am thinking these could be two separate thing. Basically, I think we'd need to support the first approach regardless of whether we support the second - so, might as well figure out what's the best way to do that.

First, to set the baseline. We could write constants to the memory using `push` operations. If we have 1K constants to write, this would take close to 3K VM cycles (1K cycles to push the value onto the stack, 1K cycles to pop them off the stack, and probably close to 1K cycles to save them to memory).

Theoretically, using the advice tape, we could speed this up significantly. The mechanism would work like this:
1. Hash the inputs outside of the VM.
2. Push the hash of the inputs onto the stack.
3. Then read the inputs from the advice tape (sing `loadw.adv` instruction) and hash them sequentially (using `rpperm` instruction) while saving the read inputs to memory.
4. Once all the inputs have been read and hashed, compare the computed hash with the hash pushed onto the stack in step 2. If the hashes are the same, we know that the memory now has the values we wanted to put there.

I've spent a little bit of time thinking about how this can be done with the current instructions, and unfortunately, practical efficiency is probably not going to be great. The main reason is that to absorb new elements into the hasher state we need to add them to the current state (rather than just replacing them). Just this step will require 8 additions to absorb 8 new elements. So, I wouldn't be surprised if we end up with needing more than 3K cycles with this approach.

But, we could add a specialized instruction to help with this task, and that could make things significantly faster. Here is one idea:

Imagine we have and instruction `pipe_adv` which does the following:
1. Reads 4 elements from the advice tape and adds them to the top 4 elements of the stack (element-wise).
2. Writes the read elements into the memory address specified by the 13th element of the stack.
3. Increments the 13th element of the stack.

The transition diagram for this instruction could look like so:
```
[C, B, A, addr, ...] -> [C + D, B, A, addr + 1, ...]
D -> mem[addr]
```
In the above:
- `C`, `B`, and `A` are words (4 elements each) on the top of the stack.
- `D` is the word read from the advice tape

Then, using this instruction, we could use the following sequence to absorb 8 elements into hasher state and to write them to memory at the same time:
```
swapw
pipe_adv
swapw
pipe_adv
rpperm
```
Thus, with 5 cycles we'd process 8 field elements, and the overall number of cycles needed to put 1K elements to memory should be slightly over 600 (or 5x improvement over the baseline).

We could go a step further and define `pipe_adv` as follows:
```
[C, B, A, addr, ...] -> [C + E, B + D, A, addr + 2, ...]
D -> mem[addr]
E -> mem[addr + 1]
```
In this case, the following sequence of instructions would process 8 elements:
```
pipe_adv
rpperm
```
And then we need under 300 cycles to move all 1K elements to memory (i.e., a 10x improvement over the baseline).

Of course, the above is a very specialized instruction and the question is whether it is worth having, or whether there are other approaches which could achieve the same result as efficiently.

**Comment by Al-Kindi-0** [comment]
Thank you @bobbinth for the nice description!
Two points, which are more of curiosities, that I have:
1.  Where does the constraint of needing to add the element, to be hashed, instead of just replacing them come from?
2. Is there a mechanism of extending the stack "indefinitely" with the constants before the start? And can such a mechanism be leveraged/modified in the context of the current issue?

**Comment by bobbinth** [comment]
> Where does the constraint of needing to add the element, to be hashed, instead of just replacing them come from?

This comes from Rescue Prime specifications. I wonder if this can be changed without the loss of security.

> Is there a mechanism of extending the stack "indefinitely" with the constants before the start? And can such a mechanism be leveraged/modified in the context of the current issue?

Yes, this is possible and, I believe, @itzmeanjan actually implemented something like this for secp256k1 point multiplication. However, this will not work in all cases - so, we still need to find an efficient for moving data from the advice tape to memory.

Another thing that comes to mind: we probably also need a revers of these operation: computing a sequential hash of some memory region. This would be useful, for example, if we want to return a lot of data as output. This data could be written to memory, then hash of this data could be computed, and then this hash could be returned via the stack.

So, we'll probably need another operation for this: maybe something like `pipe_mem`.

**Comment by Al-Kindi-0** [comment]
> This comes from Rescue Prime specifications. I wonder if this can be changed without the loss of security.

The sponge construction does allow for what is called "overwrite mode" as per section 6.2 in the reference work https://keccak.team/files/SpongeDuplex.pdf . So I don't see why it shouldn't work in our setting.
The question then is: would this simplification make your first proposal to use advice tape, without new instructions, competitive enough?

**Comment by bobbinth** [comment]
If that is possible, that would simplify things quite a bit (and not only here). I think we probably would still want specialized instructions as I think this sequential "unhashing" will be a very heavily used operation in the VM - so, we'd want to make it as efficient as possible.

**Comment by bobbinth** [comment]
A more concrete description of what we could do here. Suggestions on better names for operations/instructions are welcome.

### Add PIPE operation and related assembly instruction
As described above, the `PIPE` operation would move 8 elements from the head of the advice tape into memory via the stack. The semantics of the operation are as follows:

Denoting stack registers as $s_0, ..., s_{15}$ with $s_0$ being at the top of the stack, and advice tape items as $t_0, ..., t_7$, with $t_0$ being at the head of the tape, the operation induces the following stack transition:

$$
s_i' = s_i + t_{7 - i} \text{ for } i \in 0, ..., 7
$$

$$
s_i' = s_i  \text{ for } i \in 8, ..., 11
$$

$$
s_{12}' = s_{12} + 2
$$

To describe the above informally:
* We add 8 items at the head of the advice tape to the top 8 elements of the stack. This addition happens in "stack order" - i.e., $t_0$ is added to $s_7$, $t_1$ is added to $s_6$ etc. The added items are removed from the tape.
* We copy over the values of the following 4 stack elements to the next row.
* We increment the 13th stack element by 2.

In addition to the above, we update the memory as follows:
* $mem[s_{12}] \leftarrow \[ s_{7}' - s_{7}, s_{6}' - s_{6}, s_{5}' - s_{5}, s_{4}' - s_{4} \]$
* $mem[s_{12} + 1] \leftarrow \[ s_{3}' - s_{3}, s_{2}' - s_{2}, s_{1}' - s_{1}, s_{0}' - s_{0} \]$

Basically, we write $\[t_0, t_1, t_2, t_3\]$ into memory address specified by $s_{12}$, and $\[t_4, t_5, t_6, t_7\]$ into the memory address specified by $s_{12} + 1$. This does require two memory operations (two writes), and thus, we need to use an opcode with a low degree. I think degree 6 opcode should work here - so, we can probably use `1010100`.

The assembly instruction to invoke this operation could be `adv_pipe`.

Note: this procedure will change slightly after we migrate to RPO hash function (#474). The main difference will be that we would be able to replace field elements at the top of the stack rather than add into them.

### Add MSTREAM operation and related assembly instruction
The `MSTREAM` operation would add 8 elements from adjacent memory addresses to the elements at the top of the stack. The semantics of the operation are as follows (using the same notation as above):

$$
s_{7 - i}' = s_{7 - i} + mem[s_{12}][i] \text{ for } i \in 0, ..., 3
$$

$$
s_{3 - i}' = s_{3 - i} + mem[s_{12 + 1}][i] \text{ for } i \in 0, ..., 3
$$

$$
s_i' = s_i  \text{ for } i \in 8, ..., 11
$$

$$
s_{12}' = s_{12} + 2
$$

To describe the above informally:
* We take the word at memory address specified by $s_{12}$ and add its elements to the second word from the top of the stack.
* We take the word at memory address specified by $s_{12} + 1$ and add its elements to the word at the top of the stack.
* We copy over the values of stack elements $s_8, ..., s_{11}$ to the next row.
* We increment the 13th stack element by 2.

This requires two memory operations (two reads), and thus, we need to use an opcode with a low degree. I think degree 6 opcode should work here - so, we can probably use `1010110`.

The assembly instruction to invoke this operation could be `mem_stream`.

### Add memory advice injector decorator
The purpose of this decorator is to copy a region of memory into the advice provides and put it under the specified key. The assembly instruction for this decorator could be `adv.mem.[start addr].[num word]`.

For example, `adv.mem.4.3` would take 3 words from memory at addresses 4, 5, and 6, and would put them into the advice provider under the key specified by the top 4 elements of the stack.

To support this operation, in addition to the decorator itself, we'll need to add a method to the advice provider to support inserting into the value map. This could look something like this:
```Rust
pub fn insert_into_map(&mut self, key: Word, values: Vec<Felt>) -> Result<(), ExecutionError>
```

We could also support other variants of this decorator (though, this support could also be added later on). For example, `mem.adv` could take both the start address and number of words from the stack as well.

**Comment by bobbinth** [comment]
Closed by #486, #488, and 491.
